p2p/simulations/adapters/inproc_test.go

```diff
@@ -0,0 +1,151 @@
+// Copyright 2017 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package adapters
+
+import (
+	"bytes"
+	"encoding/binary"
+	"testing"
+	"time"
+)
+
+func TestSocketPipe(t *testing.T) {
+	c1, c2, _ := socketPipe()
+
+	done := make(chan struct{})
+
+	go func() {
+		msgs := 20
```

Comment:
These tests are only A -> B? How about A <-> B?

Reply:
I added bidirectional tests for all adapters. All connection adapters are duplex.

Why so few msgs: because 2 out of 3 adapters are using a buffer, and if you notice within those tests, I am emitting the writes, without issuing the reads (knowing that I am testing an adapter with a buffer). It is expected for the test to crash if you hit that buffer, therefore we should be careful to not emit too many messages*size.

In a real world scenario, you always have a reader and a writer, you generally never issue writes when no actor is reading.